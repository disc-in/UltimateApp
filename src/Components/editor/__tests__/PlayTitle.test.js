import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { Alert } from 'react-native';

import store from '../../../Store/testStore';
import animationSquare from '../../../Fixtures/Animation/AnimationSquare';

import ConnectedPlayTitle, { PlayTitle } from '../PlayTitle';

describe('<PlayTitle />', () => {
  afterEach(() => jest.clearAllMocks());

  const play = { animation: animationSquare, title: 'Square', uuid: '123' };
  const renamePlay = jest.fn();
  const deletePlay = jest.fn();
  const onPress = jest.fn();

  it('renders correctly', async () => {
    const { toJSON } = render(
      <Provider store={store}>
        <ConnectedPlayTitle play={play} onPress={onPress} deletePlay={deletePlay} renamePlay={renamePlay} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when safe mode', async () => {
    const { toJSON } = render(
      <Provider store={store}>
        <ConnectedPlayTitle play={play} safe />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when unsaved', async () => {
    const { toJSON } = render(
      <Provider store={store}>
        <ConnectedPlayTitle play={play} unsavedPlay />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with custom style', async () => {
    const { toJSON } = render(
      <Provider store={store}>
        <ConnectedPlayTitle play={play} style={{ backgroundColor: 'cyan' }} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('triggers onPress action', async () => {
    const { getByText } = render(
      <PlayTitle play={play} onPress={onPress} deletePlay={deletePlay} renamePlay={renamePlay} />,
    );

    fireEvent.press(getByText(play.title));

    expect(onPress).toHaveBeenCalled();
  });

  it('triggers rename action on blur', async () => {
    const { getByTestId, getByDisplayValue } = render(
      <PlayTitle play={play} onPress={onPress} customPlays={[]} deletePlay={deletePlay} renamePlay={renamePlay} />,
    );

    fireEvent.press(getByTestId('edit'));
    fireEvent(getByDisplayValue(play.title), 'onChangeText', 'New Title');
    fireEvent(getByDisplayValue('New Title'), 'onBlur');

    await expect(renamePlay).toHaveBeenCalledWith(play.uuid, 'New Title');
  });

  it('triggers rename action on validate', async () => {
    const { getByTestId, getByDisplayValue } = render(
      <PlayTitle play={play} onPress={onPress} customPlays={[]} deletePlay={deletePlay} renamePlay={renamePlay} />,
    );

    fireEvent.press(getByTestId('edit'));
    fireEvent(getByDisplayValue(play.title), 'onChangeText', 'New Title');
    await fireEvent.press(getByTestId('update'));

    await expect(renamePlay).toHaveBeenCalledWith(play.uuid, 'New Title');
  });

  it('displays errors and does not rename if new title is empty', async () => {
    const { getByTestId, getByDisplayValue, getByText } = render(
      <PlayTitle play={play} onPress={onPress} customPlays={[]} deletePlay={deletePlay} renamePlay={renamePlay} />,
    );

    fireEvent.press(getByTestId('edit'));
    fireEvent(getByDisplayValue(play.title), 'onChangeText', '');
    await fireEvent.press(getByTestId('update'));

    expect(getByText('You cannot set an empty name')).toBeDefined();
    await expect(renamePlay).not.toHaveBeenCalled();
  });

  it('displays errors and does not rename if new title is already taken', async () => {
    const existingPlay = { animation: animationSquare, title: 'Existing', uuid: '999' };
    const { getByTestId, getByDisplayValue, getByText } = render(
      <PlayTitle
        play={play}
        onPress={onPress}
        customPlays={[existingPlay]}
        deletePlay={deletePlay}
        renamePlay={renamePlay}
      />,
    );

    fireEvent.press(getByTestId('edit'));
    fireEvent(getByDisplayValue(play.title), 'onChangeText', 'Existing');
    await fireEvent.press(getByTestId('update'));

    expect(getByText('This name already exists')).toBeDefined();
    await expect(renamePlay).not.toHaveBeenCalled();
  });

  it('does not trigger delete on cancel', async () => {
    jest.spyOn(Alert, 'alert');
    const { getByTestId, getByDisplayValue } = render(
      <PlayTitle play={play} onPress={onPress} deletePlay={deletePlay} renamePlay={renamePlay} />,
    );

    fireEvent.press(getByTestId('delete'));

    // Opens confirmation Alert
    expect(Alert.alert).toHaveBeenCalled();

    // Press Cancel
    expect(Alert.alert.mock.calls[0][2][0].text).toEqual('Cancel');
    Alert.alert.mock.calls[0][2][0].onPress();

    expect(deletePlay).not.toHaveBeenCalled();
  });

  it('triggers delete on confirmation', async () => {
    jest.spyOn(Alert, 'alert');
    const { getByTestId, getByDisplayValue } = render(
      <PlayTitle play={play} onPress={onPress} deletePlay={deletePlay} renamePlay={renamePlay} />,
    );

    fireEvent.press(getByTestId('delete'));

    // Opens confirmation Alert
    expect(Alert.alert).toHaveBeenCalled();

    // Press Cancel
    expect(Alert.alert.mock.calls[0][2][1].text).toEqual('Delete');
    Alert.alert.mock.calls[0][2][1].onPress();

    expect(deletePlay).toHaveBeenCalledWith(play.uuid);
  });
});
