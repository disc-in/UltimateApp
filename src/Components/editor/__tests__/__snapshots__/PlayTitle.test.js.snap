// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayTitle /> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexBasis": "100%",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      
      Square
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="edit"
  >
    <Text />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="delete"
  >
    <Text />
  </View>
</View>
`;

exports[`<PlayTitle /> renders correctly when safe mode 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexBasis": "100%",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      
      Square
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="edit"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Object {
            "color": "#232b2b",
            "fontSize": 20,
          },
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰏫
    </Text>
  </View>
</View>
`;

exports[`<PlayTitle /> renders correctly when unsaved 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexBasis": "100%",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      * 
      Square
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="edit"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Object {
            "color": "#232b2b",
            "fontSize": 20,
          },
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰏫
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="delete"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Object {
            "color": "#232b2b",
            "fontSize": 20,
          },
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰩹
    </Text>
  </View>
</View>
`;

exports[`<PlayTitle /> renders correctly with custom style 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      Object {
        "backgroundColor": "cyan",
      },
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexBasis": "100%",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      
      Square
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="edit"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Object {
            "color": "#232b2b",
            "fontSize": 20,
          },
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰏫
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="delete"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Object {
            "color": "#232b2b",
            "fontSize": 20,
          },
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰩹
    </Text>
  </View>
</View>
`;
